File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> st = new Stack<>();
        int n = nums2.length;
        int[] nextG = new int [10001];
        for (int i = n-1;i>=0;i--){
            while(!st.isEmpty() && st.peek() <=nums2[i]){
                st.pop();
            }

            nextG[nums2[i]] = st.isEmpty() ? -1 : st.peek();
            st.push(nums2[i]);
        }

        for (int i =0; i<nums1.length;i++){

            nums1[i] = nextG[nums1[i]];
        }

        return nums1;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st=[]
        arr=[-1]*len(nums2)

        for i in range(len(nums2)):
            while st and nums2[st[-1]]< nums2[i]:

                x=st.pop()
                arr[x]=nums2[i]
            st.append(i)

        ge={}

        for i in range(len(arr)):
            ge[nums2[i]]=arr[i]
        
        res=[]

        for num in nums1:
            res.append(ge[num])

        return res